File: code/raytracer/python/resources.py
import brdfs
from utils import jsonfile2obj

# The resources json file has the name of the brdfs, this methods associates the actual function instead


class Resources(object):
    def __init__(self, materials):
        self.materials = {m.name: m for m in materials}

    def get_brdf_materials(self, material_names):
        return [self.materials[m] for m in material_names if isinstance(self.materials[m], BRDFMaterial)]

    def get_reflective_materials(self, material_names):
        return [self.materials[m] for m in material_names if isinstance(self.materials[m], ReflectiveMaterial)]

    def get_ambient_materials(self, material_names):
        return [self.materials[m] for m in material_names if isinstance(self.materials[m], BRDFMaterial) and
                self.materials[m].use_for_ambient]


class BRDFMaterial(object):
    def __init__(self, name, color, brdfParams, brdf, use_for_ambient):
        self.name = name
        self.color = color
        self.brdfParams = brdfParams
        self.brdf = brdf
        self.use_for_ambient = use_for_ambient


class ReflectiveMaterial(object):
    def __init__(self, name, reflectivity):
        self.name = name
        self.reflectivity = reflectivity


def loadResources(resourcesFile):
    return jsonfile2obj(resourcesFile, object_hook=resourcesHook)


def resourcesHook(obj):
    if '__type__' in obj:
        if obj['__type__'] == "brdf_material":
            brdfFunc = getattr(brdfs, obj['brdf'])
            use_for_ambient = False
            if 'use_for_ambient' in obj:
                use_for_ambient = obj['use_for_ambient']
            return BRDFMaterial(obj['name'], obj['color'], obj['brdfParams'], brdfFunc, use_for_ambient)
        if obj['__type__'] == "reflective_material":
            return ReflectiveMaterial(obj['name'], obj['reflectivity'])
        if obj['__type__'] == "resources":
            return Resources(obj['materials'])
    return obj

File: code/raytracer/javascript/resources.js
//The resources json file has the name of the brdfs, this methods associates the actual function instead
function loadResources(resources)
{
	var materialDict = {};
	for(var i=0; i<resources.materials.length; i++)
	{
		materialDict[resources.materials[i].name] = resources.materials[i];
	}
	resources.materials = materialDict; 
	for(var materialName in resources.materials)
	{
		if(resources.materials[materialName].__type__ == "brdf_material")
		{ 
			var brdf_name = resources.materials[materialName].brdf;
			var functPtr = window[brdf_name];
			resources.materials[materialName].brdf = functPtr;
		}
	}

	resources.getBrdfMaterials = function(materialNames){
		var brdfMaterials = [];
		for(var i=0; i<materialNames.length; i++)
		{
			if(this.materials[materialNames[i]].__type__ == "brdf_material")
				brdfMaterials.push(this.materials[materialNames[i]]);
		}
		return brdfMaterials;
	};
	resources.getReflectiveMaterials = function(materialNames){
		var reflectiveMaterials = [];
		for(var i=0; i<materialNames.length; i++)
		{
			if(this.materials[materialNames[i]].__type__ == "reflective_material")
				reflectiveMaterials.push(this.materials[materialNames[i]]);
		}
		return reflectiveMaterials;
	};
	resources.getAmbientMaterials = function(materialNames){
		var ambientMaterials = [];
		for(var i=0; i<materialNames.length; i++)
		{
			if(this.materials[materialNames[i]].use_for_ambient)
				ambientMaterials.push(this.materials[materialNames[i]]);
		}
		return ambientMaterials;
	}
}